File: QID-3385-SFEhillquantile/SFEhillquantile.m
%% clear all variables
clear all
close all
clc

disp('Please input Quantile q, and Excess k as [0.01, 10]');
disp(' ') ;
para = input('[q, k]=');
while length(para) < 2
  disp('Not enough input arguments. Please input in 1*2 vector form like [0.01, 10]');
  disp(' ') ;
  para = input('[q, k]=');
end
q = para(1);              % quantile
k = para(2);              % excess

if(k < 1 && k> n)
    error('SFEhillquantile: excess should be greater than 0 and less or equal than the number of elements.')
end
if(q < 0 && q> 1)
    error('SFEhillquantile: please give a rational quantile value.')
end

x    = gprnd(0.1,1,0,100,1); % generalized pareto random numbers
x    = sort(x);
n    = size(x,1);
rest = hillgp1(x,k);  % ML-estimation of gamma 
xest = x(k)+x(k)*(((n/k)*(1-q))^(-rest)-1) % Hill-quantil estimation

File: QID-3385-SFEhillquantile/SFEhillquantile.R
# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

# install and load packages
libraries = c("fExtremes")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
  install.packages(x)
})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

# Main computation
x = rgpd(100, xi = 0.1)  # generate a generalized pareto distributed variables 
x = sort(x)
n = length(x)
q = "0.01"
k = "10"

message = "      Quantile"
default = q
q = winDialogString(message, default)
q = type.convert(q, na.strings = "NA", as.is = FALSE, dec = ".")

message = "      Excess"
default = k
k = winDialogString(message, default)
k = type.convert(k, na.strings = "NA", as.is = FALSE, dec = ".")

if (k < 8 && k > (n - 1)) warning("SFEhillquantile: excess should be greater than 8 and less than the number of elements.", 
                                  call. = FALSE)
if (q < 0 && q > 1) warning("SFEhillquantile: please give a rational quantile value.", 
                            call. = FALSE)

rest = gpdFit(x, nextremes = k, type = "mle")				# ML-estimation of gamma 
rest = rest@parameter

# the Hill-quantile value
(xest = x[k] + x[k] * (((n/k) * (1 - q))^(-rest$u) - 1))	# Hill-quantil estimation